crates/wasi-common/src/virtfs.rs

```diff
@@ -255,7 +257,14 @@ impl Handle for InMemoryFile {
     fn read_vectored(&self, iovs: &mut [io::IoSliceMut]) -> Result<usize> {
         trace!("read_vectored(iovs={:?})", iovs);
         trace!("     | *read_start={:?}", self.cursor.get());
-        self.data.borrow_mut().preadv(iovs, self.cursor.get())
+        let read = self.data.borrow_mut().preadv(iovs, self.cursor.get())?;
+        let update = self
+            .cursor
+            .get()
+            .checked_add(read.try_into()?)
+            .ok_or(Errno::Fault)?;
```

Comment:
Looking at POSIX, this looks like more of an `EINVAL` situation than `EFAULT`.

https://pubs.opengroup.org/onlinepubs/009695399/functions/pread.html

`EINVAL` is for "the offset argument is not valid".

Reply:
Thank you for the pointer; I've updated it to using `EINVAL`.